[Multilane] Fixes PR #7325.

agalbachicar opened the pull request:
After PR #7325 was merged, it is necessary to roll back some modifications made during the review of the PR. Here is a summary:

- `Connection`'s constructors retain their signature instead of constructing structures to group them.
- Tolerance magnitude conversion in GTest idioms to compare Endpoint types is reverted as well as the TODOs.
- `EndpointZ::reverse` is buggy. Not only  code is corrected but a new doc string on the method is introduced trying to avoid future confusions.
- `Group`'s doc strings with reference to ID uniqueness is removed.

<!-- Reviewable:start -->
---
This change is [<img src="https://reviewable.io/review_button.svg" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/robotlocomotion/drake/7380)
<!-- Reviewable:end -->

agalbachicar commented:
+@maddog-tri for feature review please.

---

Review status: 0 of 5 files reviewed at latest revision, all discussions resolved.

---



*Comments from [Reviewable](https://reviewable.io:443/reviews/robotlocomotion/drake/7380#-:-KxnnEBAFYd8Z8DMaq9g:b-5klkw7)*
<!-- Sent from Reviewable.io -->

maddog-tri commented:
Reviewed --- please take a look.  

---

Reviewed 5 of 5 files at r1.
Review status: all files reviewed at latest revision, 4 unresolved discussions.

---

*[a discussion](https://reviewable.io:443/reviews/robotlocomotion/drake/7380#-Kxo83r7XIdCqDLXK5Kd:-Kxo83r8T7ryOHaWzDOx:b-ws0cf8) (no related file):*
Oh, ha, HA!  There _is_ a bug in the original code (and thus in monolane, too)... but it's not with `theta`, it's with `theta_dot`!   (Calling @sherm1 to enjoy the laugh, too.) 

I realized this when I was suggesting a better docstring addition for `EndpointZ::reverse()`.

The "reverse" operation basically flips the signs of `s` and `r` and thus affects derivatives with respect to ds and dr.  So:
  * `z_dot` = dz/ds ---> flipped to `-z_dot`;
  * `theta` = arctan(dz/dr) ---> flipped `-theta`;
  * `theta_dot` = d(theta)/ds ---> flipped _twice_, so it should remain `theta_dot`!

Could you do the one-character fix to `monolane/builder.h` in this PR, too?

Also, I apologize for thinking you scribbled on a unit-test that was copied over from a carefully-crafted unit-test in `monolane` --- I see now that `monolane` had no specific testing of `theta` and `theta_dot`.

---

*[a discussion](https://reviewable.io:443/reviews/robotlocomotion/drake/7380#-KxoCpiE1YauVR_FVfYk:-KxoCpiFbJu89kfyPewF:b-n0hujv) (no related file):*
Please add a TODO to `multilane_builder_test.cc` along the lines of:
```
// TODO(maddog-tri)  Test use of Endpoint::reverse() with non-zero theta and theta_dot,
//                   checking that the orientation of the resulting road surface is continuous,
//                   off the centerline, at the branch-point where two connections connect.
```

---

*[drake/automotive/maliput/multilane/connection.h, line 82 at r1](https://reviewable.io:443/reviews/robotlocomotion/drake/7380#-Kxo4RY13mvBNNGyhtDf:-Kxo4RY13mvBNNGyhtDg:b6hs0cp) ([raw file](https://github.com/robotlocomotion/drake/blob/ce149c286bcbe2c7fc09c32d4c424409eb2f408e/drake/automotive/maliput/multilane/connection.h#L82)):*
> ```objc
>   /// As theta and theta_dot are part of the superelevation of the road, when
>   /// going in the opposite direction there is a flip of the axis. Consequently,
>   /// z_dot, theta and theta_dot must be negative.
> ```

This is a confusing addition (esp. the "must be negative", which sounds like a precondition on the input).

Here is a suggested alternative:
```
/// Reversing direction is equivalent to rotating s (and along with it, r) 
/// around the h-axis by 180 degrees, thus flipping the signs of z_dot
/// and theta.
```

---

*[drake/automotive/maliput/multilane/connection.h, line 84 at r1](https://reviewable.io:443/reviews/robotlocomotion/drake/7380#-KxoCd0RBilwVw-9xnfF:-KxoCd0RBilwVw-9xnfG:b735dst) ([raw file](https://github.com/robotlocomotion/drake/blob/ce149c286bcbe2c7fc09c32d4c424409eb2f408e/drake/automotive/maliput/multilane/connection.h#L84)):*
> ```objc
>   /// z_dot, theta and theta_dot must be negative.
>   EndpointZ reverse() const {
>     return EndpointZ(z_, -z_dot_, -theta_, -theta_dot_);
> ```

Per side-thread above:  oops, this should be `theta_dot_`, not `-theta_dot_`.  (Really, it should. :) )

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/robotlocomotion/drake/7380#-:-KxoEQKEETdkzd5fz7L-:b-pyn1sa)*
<!-- Sent from Reviewable.io -->

sherm1 commented:
*[a discussion](https://reviewable.io:443/reviews/robotlocomotion/drake/7380#-Kxo83r7XIdCqDLXK5Kd:-KxoFqgM5D8yS6VfiQ_i:b-4xfayr) (no related file):*
<details><summary><i>Previously, maddog-tri (Matt Marjanović) wrote…</i></summary><blockquote>

Oh, ha, HA!  There _is_ a bug in the original code (and thus in monolane, too)... but it's not with `theta`, it's with `theta_dot`!   (Calling @sherm1 to enjoy the laugh, too.) 

I realized this when I was suggesting a better docstring addition for `EndpointZ::reverse()`.

The "reverse" operation basically flips the signs of `s` and `r` and thus affects derivatives with respect to ds and dr.  So:
  * `z_dot` = dz/ds ---> flipped to `-z_dot`;
  * `theta` = arctan(dz/dr) ---> flipped `-theta`;
  * `theta_dot` = d(theta)/ds ---> flipped _twice_, so it should remain `theta_dot`!

Could you do the one-character fix to `monolane/builder.h` in this PR, too?

Also, I apologize for thinking you scribbled on a unit-test that was copied over from a carefully-crafted unit-test in `monolane` --- I see now that `monolane` had no specific testing of `theta` and `theta_dot`.
</blockquote></details>

Ha! Fun. Thanks @maddog-tri. I'm glad my bad feeling about that code uncovered a bug -- too bad I wasn't smart enough to identify the real problem!

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/robotlocomotion/drake/7380)*
<!-- Sent from Reviewable.io -->

agalbachicar commented:
I've addressed all the comments. PTAL.

---

Review status: 3 of 7 files reviewed at latest revision, 4 unresolved discussions.

---

*[a discussion](https://reviewable.io:443/reviews/robotlocomotion/drake/7380#-Kxo83r7XIdCqDLXK5Kd:-KxoINLfScxqfSantzTs:b-xfhi8a) (no related file):*
<details><summary><i>Previously, sherm1 (Michael Sherman) wrote…</i></summary><blockquote>

Ha! Fun. Thanks @maddog-tri. I'm glad my bad feeling about that code uncovered a bug -- too bad I wasn't smart enough to identify the real problem!
</blockquote></details>

Done. No problem at all! The original intention of these PRs is to provide the necessary API changes as well as test coverage.

---

*[a discussion](https://reviewable.io:443/reviews/robotlocomotion/drake/7380#-KxoCpiE1YauVR_FVfYk:-KxoItez8eGX-Y7eP73x:b-896fix) (no related file):*
<details><summary><i>Previously, maddog-tri (Matt Marjanović) wrote…</i></summary><blockquote>

Please add a TODO to `multilane_builder_test.cc` along the lines of:
```
// TODO(maddog-tri)  Test use of Endpoint::reverse() with non-zero theta and theta_dot,
//                   checking that the orientation of the resulting road surface is continuous,
//                   off the centerline, at the branch-point where two connections connect.
```
</blockquote></details>

Done.

---

*[drake/automotive/maliput/multilane/connection.h, line 82 at r1](https://reviewable.io:443/reviews/robotlocomotion/drake/7380#-Kxo4RY13mvBNNGyhtDf:-KxoGQ9lfO_9NCobY5wE:b-896fix) ([raw file](https://github.com/robotlocomotion/drake/blob/ce149c286bcbe2c7fc09c32d4c424409eb2f408e/drake/automotive/maliput/multilane/connection.h#L82)):*
<details><summary><i>Previously, maddog-tri (Matt Marjanović) wrote…</i></summary><blockquote>

This is a confusing addition (esp. the "must be negative", which sounds like a precondition on the input).

Here is a suggested alternative:
```
/// Reversing direction is equivalent to rotating s (and along with it, r) 
/// around the h-axis by 180 degrees, thus flipping the signs of z_dot
/// and theta.
```
</blockquote></details>

Done.

---

*[drake/automotive/maliput/multilane/connection.h, line 84 at r1](https://reviewable.io:443/reviews/robotlocomotion/drake/7380#-KxoCd0RBilwVw-9xnfF:-KxoGYUHNhWpdL3CS5B5:b-896fix) ([raw file](https://github.com/robotlocomotion/drake/blob/ce149c286bcbe2c7fc09c32d4c424409eb2f408e/drake/automotive/maliput/multilane/connection.h#L84)):*
<details><summary><i>Previously, maddog-tri (Matt Marjanović) wrote…</i></summary><blockquote>

Per side-thread above:  oops, this should be `theta_dot_`, not `-theta_dot_`.  (Really, it should. :) )
</blockquote></details>

Done.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/robotlocomotion/drake/7380#-:-KxoIv1B1H9wa29-BsRi:b76ci8)*
<!-- Sent from Reviewable.io -->

maddog-tri commented:
<img class="emoji" title=":lgtm:" alt=":lgtm:" align="absmiddle" src="https://reviewable.io/lgtm.png" height="20" width="61"/>  --- and don't forget to amend the commit message appropriately.

---

Reviewed 4 of 4 files at r2.
Review status: all files reviewed at latest revision, 1 unresolved discussion.

---



*Comments from [Reviewable](https://reviewable.io:443/reviews/robotlocomotion/drake/7380#-:-KxoMXgoSHYykl2cSAod:biv6ah7)*
<!-- Sent from Reviewable.io -->

agalbachicar commented:
+@ggould-tri for platform review please.

---







*Comments from [Reviewable](https://reviewable.io:443/reviews/robotlocomotion/drake/7380#-:-KxrjHziK72iF-FLwNwR:b-irfx5k)*
<!-- Sent from Reviewable.io -->

ggould-tri commented:
Reviewed 3 of 5 files at r1, 4 of 4 files at r2.
Review status: all files reviewed at latest revision, 1 unresolved discussion.

---



*Comments from [Reviewable](https://reviewable.io:443/reviews/robotlocomotion/drake/7380)*
<!-- Sent from Reviewable.io -->

ggould-tri commented:
<img class="emoji" title=":lgtm:" alt=":lgtm:" align="absmiddle" src="https://reviewable.io/lgtm.png" height="20" width="61"/>

---

Review status: all files reviewed at latest revision, 1 unresolved discussion.

---



*Comments from [Reviewable](https://reviewable.io:443/reviews/robotlocomotion/drake/7380#-:-KxroVrRUT-6TtZdXzAR:bnfp4nl)*
<!-- Sent from Reviewable.io -->
